Improved And Standardized

XedinUnknown opened the pull request:
- Added and improved documentation;
- Split topical logic across multiple abstract classes;
- Added exception;
- Fixed standards.

Depends on #5.

XedinUnknown committed:
Centralized and abstracted the adding of the hook itself (#5)

XedinUnknown committed:
Amended test to expect manager method (#5)

This is needed because `add_filter()` is mocked by WP Mock itself,
and therefore cannot have expectations, while expectations are necessary.
WP Mock's hook expectations expect exact values for parameters, so it is
impossible to expect certain types. This made proper testing impossible,
since `attach()` is not guaranteed to hook the very same callable that was
passed to it(); only that the callable will get invoked.

XedinUnknown committed:
Added test for `_addHook()`

XedinUnknown committed:
Centralized and normalized the running of WP handlers (#5)

XedinUnknown committed:
Added tests for event triggering (#5)

XedinUnknown committed:
`trigger()` now passes an event object to handlers (#5)

Also changed normalization logic, which now accepts more args.
Finally, added event factory.

XedinUnknown committed:
Trigger test amended (#5)

XedinUnknown committed:
All events invocations normalized to same signature (#5)

All event handlers, regardless of whether they are invoked by the manager
or by WordPress, now receive one parameter: the event instance.

XedinUnknown committed:
Fixed expecting callback byref (#5)

XedinUnknown committed:
Added test for creating callback wrappers (#5)

XedinUnknown committed:
Fixed bug with same handler for different events creating new wrappers (#5)

XedinUnknown committed:
Added detail to callback wrapper creation test (#5)

Now also tests consistent receiving of event arg, and stopping of event
propagation.

XedinUnknown committed:
Added test for callback wrapper cache (#5)

XedinUnknown committed:
Added exception interface and classes (#6)

XedinUnknown committed:
Normalized and split logic, corrected and added docs, fixed standards (#6)

XedinUnknown committed:
Updated tests (6#)

XedinUnknown committed:
Fixed incompatible method declarations (#6)

XedinUnknown committed:
Code standards fixed (#6)

mecha commented on src/AbstractNormalizedEventManager.php:
```diff
@@ -0,0 +1,77 @@
+                \call_user_func_array($callback, array($event));
+            }
+
+            return $event;
```
I love how you handled this using closure returns. And support for the propagation flag is nice bonus. Truly amazing! :)

If I may add however, this won't allow the manager to be used for WordPress filters.

As an example, let's say I want to remove the _Quick Edit_ row action from all posts. The `post_row_actions` WordPress filter can be used to return an array of row actions. If I use an event manager that uses this closure as a wrapper for my handler, the `Event` instance will be returned, even if my handler complies to standard and depends on a `$event` argument.

Consider:

```php
$manager->attach('post_row_actions', function($event) {
    // Get params array and then the row actions by reference
    $params     = $event->getParams();
    $rowActions = &$params[0];
    // Remove "Quick Edit"
    unset($rowActions['inline hide-if-no-js']);
    // Set new params
    $event->setParams($params);
}
```

The above adheres to the standard. Action has been taken in the handler and the `Event`'s data has changed. This is all allowed. However, WordPress requires a _result_, and currently the wrapper only returns the _Event_ instance.

Since we cannot count on WordPress adhering to the PSR-14 event manager standard, if we wish to make the event manager compatible with attaching handlers to WordPress hooks, I recommend the following changes inside the wrapper:

Detect if the event is a WordPress event. This is easy - WordPress-triggered hooks do not pass `Event` instances.

```php
$isWpEvent = !\count($args) || !$args[0] instanceof EventInterface;
$event     = $isWpEvent
    ? $me->createEvent($name, $args)
    : $args[0];
```

So far, nothing has changed except for the addition of the flag `$isWpEvent`. This can now be used at `return`-time:

```
return $isWpEvent
    ? $event->getParam(0)
    : $event;
```

Since WordPress treats the first-argument as the "value to be modified by filters", if a WordPress event is detected, the first event parameter is returned instead of the `Event` instance. A more standardized way would be to settle on a param "key", such as `"result"` or perhaps something less intrusive like `"_filter"`.

This is of course like I said, if we wish to make the event manager capable of attaching handlers to WordPress-triggered hooks.

XedinUnknown replied:
Thank you for your kind words. And also for the explanation of how this breaks WP. I understand now, and completely agree that we should be compatible with WP. After all, that's the point.
However, what troubles me is that if we return the parameter, then the next handler will not receive the same event instance. Of course, a new instance will be created. But the data will not be retained. Propagation settings will not have effect and every handler will get unmodified data.

XedinUnknown replied:
If we make an assumption that hooks obey the two below conditions, it could be possible to create a context for a hook and store it in the manager. Every time a handler is invoked, it could then return the actual data instead of the event, and populate the context with it, subsequently restoring event data from that context. The context is a sort of a storage of the current state of the hook handler input. But how safe are these assumptions? I'm particularly concerned about 2.
1. Cannot be called recursively.
I.e. A handler of `the_title` cannot do `apply_filters('the_title')`.
2. Cannot be run twice.
E.g. two places in the code that do `apply_filters('the_title')`.

mecha replied:
I don't think (2) can be assumed. Consider WPRA F2P. When it creates posts it triggers `the_title`, if I recall correctly. But of course this isn't perhaps a valid example, since it is only a problem if a filter is triggered twice in the same runtime process. My point is that there are many instances where plugins or themes do WordPress-like things and utilize WordPress filters to make their code compatible with other sources.

XedinUnknown replied:
Another way could be: for each action, hook in a very very very late handler, which transforms return value. The priority could be `PHP_INT_MAX`. This could allow us to have something do a last modification, even if we can't know which one is last from inside the wrapper.

mecha replied:
Interesting. What if instead of modifying the return value on a very late hook, we clear the cached `EventInterface` instance?

Every handler wrapper closure can attempt to fetch the event instance. So if a handler is hooked into `the_title`, the wrapper would attempt `$me->_getCachedEvent("the_title")`. If it does not exist, it gets created.

This event instance would be retrieved by every wrapper, so we'd elimintate the need to pass it along as a paramter or as the context.

Then on the late hook, this cached instance can be reset, so that further triggers of the same hook won't fetch this instance.

XedinUnknown replied:
WRT fetching cached event data: exactly my thoughts.
WRT clearing the event: now I understand, thanks. Sounds like this could work.

mecha replied:
## Summary of Voice Call

The combined solution involves using a cache of `EventInterface` instances, mapped by the event name. Each event will not store its corresponding `EventInterface` instance in the cache for the duration of the event chain.

These cached instances will be created on demand: when a handler _wrapper_ attempts to get the event instance, the cache will create it if it doesn't exist and will also register a handler with priority `PHP_INT_MAX`.

This (semi-)ensures that the handler will be called last in the chain. It's purpose is simple: it will delete the event instance from the cache and unhook itself. This is so that if an event with the same name is later triggered, it receives a new event instance. In these cases, a new late handler will be registered with priority `PHP_INT_MAX` that will do the same thing.

All of this is to ensure that every chain of events gets the its own `EventInterface` instance that persists until the chain ends.

----

There was a bug in the [`AbstractWpEventManager#_detach`][1] where the priority argument was not being passed to `remove_filter`. I will be pushing a fix for this shortly.

[1]: https://github.com/Dhii/wp-events/blob/task/bring-up-to-dhii-standard_6/src/AbstractWpEventManager.php#L102

mecha commented on src/AbstractEventManager.php:
```diff
@@ -0,0 +1,137 @@
+     *
+     * @return array The result of merging the arrays.
+     */
+    protected function _mergeArgs($base, $other)
```
Forgive me if this is a stupid question, but is there any particular reason for not using `array_merge()`?

XedinUnknown replied:
Yes. It will not overwrite numeric keys.

mecha replied:
Ah, of course!

mecha commented on src/AbstractWrapperCachingEventManager.php:
```diff
@@ -0,0 +1,142 @@
+    {
+        $itemHashes = array();
+        foreach ($array as $_idx => $_item) {
+            if (\is_array($_item)) {
```
Consider using [`gettype()`][1] and a `switch` statement with type class constants for switch cases.

[1]: http://php.net/manual/en/function.gettype.php

mecha reviewed (changes requested):
All in all I love what you did here. Great work :)

If we can at least _discuss_ usage with WordPress triggered hooks, that would be great. It would allow plugins to use the event manager for everything - be it WordPress hooks or their own custom events. The uniformity and consistency would be amazing.

mecha committed:
Fixed priority argument not being passed through (#6)

mecha committed:
Added event instance cache (#6)

mecha committed:
Handler wrapper uses cached instances (#6)

mecha committed:
Added func test for `AbstractNormalizedEventManager` (#6)

mecha committed:
Removed public `createEvent` method (#6)

mecha committed:
Added priority to public `detach()` method (#6)

mecha committed:
Fixed detaching of wrapper handlers (#6)

mecha committed:
Added public event cache getter method (#6)

This is unfortunately the only solution for PHP 5.3.
Variables in closures cannot call protected methods, even if
they are instances of that class.

mecha committed:
Added public event cache removal method (#6)

This is unfortunately the only solution for PHP 5.3.
Variables in closures cannot call protected methods, even if
they are instances of that class.

XedinUnknown commented on src/AbstractNormalizedEventManager.php:
```diff
@@ -13,6 +13,150 @@
+
+        $callback = function($value) use ($me, $event, &$callback, $priority) {
+            $me->_zRemoveCachedEvent($event->getName());
+            $me->_detach($event, $callback, $priority);
```
This doesn't look like it will work on PHP 5.3. See previous comment suggestion, too.

XedinUnknown commented on src/AbstractWpEventManager.php:
```diff
@@ -99,7 +99,7 @@ protected function _clearListeners($event)
     {
         $eventObject = $this->_normalizeEvent($event);
-        \remove_filter($eventObject->getName(), $callback);
+        \remove_filter($eventObject->getName(), $callback, $priority);
```
Good catch! :)

XedinUnknown commented on src/AbstractWrapperCachingEventManager.php:
```diff
@@ -37,6 +37,22 @@ protected function _getHandlerWrapper($name, $callback)
+     *
+     * @return AbstractWpEventManager This instance.
+     */
+    protected function _detach($event, $callback, $priority = self::DEFAULT_PRIORITY)
```
Excellent, fixes the inability to detach handlers. Good use of `_getHandlerWrapper()`.

XedinUnknown commented on src/AbstractNormalizedEventManager.php:
```diff
@@ -13,6 +13,150 @@
 abstract class AbstractNormalizedEventManager extends AbstractWpEventManager
```
The functionality that now facilitates WP compat is complicated, and IMHO deserves an extended explanation in class docs, or somewhere else.

XedinUnknown commented on src/AbstractNormalizedEventManager.php:
```diff
@@ -13,6 +13,150 @@
+     *
+     * @return $this This instance.
+     */
+    protected function _registerCacheClearHandler(EventInterface $event)
```
There doesn't appear to be a WP Mock expectation of a filter being removed, so I guess we cannot really test `_detach()`. But what we can test for is whether the cache-clear handler returned by this method _calls_ `_detach()`. See this [`testAttach()`](https://github.com/Dhii/wp-events/blob/task/bring-up-to-dhii-standard_6/test/functional/EventManagerTest.php#L76).

XedinUnknown reviewed (commented):
Overall, love it! It's amazing how using cache it's possible to actually unhook the wrapper by being given the handler. And that it was possible to actually **standardize** something WP for once. I think that many people will be grateful for this standardization, and for being able to change _all_ event parameters, not to mention that the number-of-args param is now _gone_. I know _I'm_ grateful. Why we so naisse?

mecha committed:
Added abstract public event detach method (#6)

This is unfortunately the only solution for PHP 5.3.
Variables in closures cannot call protected methods, even if
they are instances of that class.

mecha committed:
Separated clear cache handler creation and registration (#6)

mecha committed:
Renamed clear cache handler methods (#6)

mecha committed:
`_registerClearCacheHandler()` returns handler callback (#6)

mecha committed:
Fixed undefined `$priority` variable (#6)

mecha committed:
Added function tests for clear cache handler (#6)

mecha committed:
Added functional test for `detach()` (#6)

XedinUnknown commented on src/AbstractNormalizedEventManager.php:
```diff
@@ -32,6 +32,17 @@
+     *
+     * @return bool True on success, false on failure
+     */
+    abstract public function detach($event, $callback, $priority = self::DEFAULT_PRIORITY);
```
Why `abstract public` instead of `public function _zDetach()`? Descendants of this class are not guaranteed to implement an interface exposing `detach()`, and therefore should not be forced to do this. Only you know the use-case of the descendants.

mecha replied:
Fixed

XedinUnknown commented on src/AbstractNormalizedEventManager.php:
```diff
@@ -139,21 +150,38 @@ public function _zRemoveCachedEvent($name)
         $callback = function($value) use ($me, $event, &$callback, $priority) {
             $me->_zRemoveCachedEvent($event->getName());
-            $me->_detach($event, $callback, $priority);
+            $me->detach($event, $callback, $priority);
```
See [this comment](https://github.com/Dhii/wp-events/pull/10/files/4c0b8e7ab21ecb2d4eecd234aec6d177727acf6a..a8480673207370a699a71a74643ccbfff7663ae2#r103468882).

mecha replied:
Fixed

mecha committed:
Abstract `detach()` is now public `_zDetach()` (#6)

XedinUnknown commented on src/AbstractNormalizedEventManager.php:
```diff
@@ -40,7 +40,12 @@
      * @return bool True on success, false on failure
      */
-    abstract public function detach($event, $callback, $priority = self::DEFAULT_PRIORITY);
+    public function _zdetach($event, $callback, $priority = self::DEFAULT_PRIORITY)
```
Case of letter `D` is wrong.

mecha committed:
Fixed method name case error (#6)

XedinUnknown commented:
Needs to add Mockery dependency with a version constraint. Because of WP Mock, we are not using the latest version of Mockery; and yet, we are still using it, and consuming its interface.

XedinUnknown commented:
So, after giving the code some thorough thought, it appears now that `stopPropagation()` will prevent only the hooks registered with the manager from running. Hooks registered with the WP native API will still run.

mecha commented:
@XedinUnknown See #11 and #12

mecha commented:
@XedinUnknown Also, what Mockery version should we depend on?

*Edit: It was decided to depend on `^0.9.5`*

mecha committed:
Added version constraint for Mockery

mecha committed:
Fixed project include path

mecha commented:
@XedinUnknown This looks ready for merging to me. Do you approve?
